finality-aleph/src/base_protocol/handler.rs

```diff
@@ -1,45 +1,147 @@
-use std::collections::{HashMap, HashSet};
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::{Display, Formatter, Result as FmtResult},
+};
 
 use parity_scale_codec::{DecodeAll, Error as CodecError};
-use sc_network::{config::FullNetworkConfiguration, PeerId};
+use sc_network::{config::NetworkConfiguration, service::traits::Direction, PeerId};
 use sc_network_common::{role::Roles, sync::message::BlockAnnouncesHandshake};
 use sp_runtime::traits::{Block, Header, Saturating};
 
 use crate::{BlockHash, BlockNumber};
 
+#[derive(Clone, Debug)]
+pub enum DisconnectError {
+    PeerWasNotConnected,
+}
+
+/// The role of the connected node.
+#[derive(Clone, Copy, Debug)]
+pub enum Role {
+    /// A full node, the expected type.
+    Full,
+    /// A light node, we support these connecting to us, but don't provide any implementations.
+    Light,
+}
+
+impl From<Roles> for Role {
+    fn from(roles: Roles) -> Self {
+        match roles.is_full() {
+            true => Role::Full,
+            false => Role::Light,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct PeerInfo {
+    role: Role,
+    direction: Direction,
+}
+
+impl PeerInfo {
+    pub fn new(role: Role, direction: Direction) -> Self {
+        PeerInfo { role, direction }
+    }
+}
+
+/// Reasons to refuse connecting to a peer.
+#[derive(Clone, Debug)]
 pub enum ConnectError {
+    /// We weren't able to decode the handshake.
     BadlyEncodedHandshake(CodecError),
-    BadHandshakeGenesis,
-    PeerAlreadyConnected,
-    TooManyFullInboundPeers,
-    TooManyFullOutboundPeers,
-    TooManyLightPeers,
+    /// The peer is running on a different chain.
+    BadHandshakeGenesis(BlockHash),
+    /// The peer is already connected.
+    AlreadyConnected(PeerId),
+    /// There are too many peer of this type already connected.
+    TooManyPeers(Role, Direction),
```

Comment:
In theory it allows creation of pair `Light, Outbound` which doesn't make sense, although reducing the number of variants is cool. Consider a meet-in-the-middle alternative with 2 variants instead of 3:
```rust
TooManyFullPeers(Direction),
TooManyLightPeers,
```
This is minor stuff tho, feel free to ignore & resolve according to your preference

Reply:
Yeah, good point, it bothered me a bit as well. Fixed as suggested.